### PR TITLE
feat(dashboard): visualize semantic agent traces

### PR DESCRIPTION
**Pull request by hallerite**
## Summary

- add a third **Semantic** trace view alongside Transcript and Timeline
- project persisted `MessageNode.semantic_parents` into a causal agent/context graph while keeping the physical training-message DAG unchanged
- center the root agent, reuse bounded subagent lanes, split fork/join ports, and render compactions as transitions into new context segments
- fold long uninterrupted runs of ordinary model calls while preserving every semantic endpoint, context boundary, active call, and selected call
- add exact-time hover details, an on-demand call inspector, direct navigation to a highlighted transcript entry, and context/agent token summaries
- recover display-only continuations when the physical message chain proves a completed tool round-trip, while labeling ambiguous fragments as unlinked
- page Semantic episode navigation across unloaded result pages
- keep traces without semantic metadata fully compatible

## Views

- **Transcript** remains the complete message and tool transcript.
- **Timeline** remains the wall-clock view of physical prefix branches.
- **Semantic** is a top-to-bottom causal view of agent execution. `continuation` keeps calls in one context, `subagent_call` and `subagent_return` describe delegation, and `compaction` starts a new context for the same agent. Unknown edge labels remain visible without inventing semantics.

Semantic uses causal rank rather than elapsed time for vertical placement; exact timing stays in hover details and Timeline. Child positions are reused after return, so width reflects peak concurrency rather than the total number of subagents created during a rollout. Return edges use stable join rails near the consuming parent to keep concurrent joins legible.

Ordinary-call runs retain two calls of context at each end and collapse only their unstructured middle. Expanding a run preserves scroll position, and selecting a call temporarily reveals it before the compact layout is restored.

Episode opens are version-guarded across both episode and timeline fetches, so rapid navigation cannot let a stale Semantic fallback overwrite the newer episode's view. In-place rerenders preserve the viewport, while closing and reopening resets its identity and recenters the root—even for the same episode.

Compaction styling is derived from the complete incoming edge set, so a context remains visibly compacted when its first node also has return or custom parents.

The viewer does not mutate persisted lineage. When cancellation leaves a completed model call without its final semantic continuation, it rejoins that call only if the physical parent path consists exclusively of tool results belonging to the preceding model call. The inspector marks the relationship as recovered from physical tool flow. Anything less certain remains visible as an explicitly unlinked fragment instead of being presented as a new context.

## Real rollout examples

These are completed RedSearch rollouts on DeepSeek V4 Flash with concurrent subagents and automatic compaction. None contains the 300-second subagent timeout marker.

**Trace 3 — one completed subagent round-trip and one compaction**

Semantic trace 3

**Trace 6 — four completed subagent round-trips and three compactions**

Semantic trace 6

**Trace 15 — five completed subagent round-trips and two compactions**

Semantic trace 15

## Validation

- `uv run pytest tests/unit/test_dashboard_timeline.py`
- `uv run pre-commit run --all-files`
- `node --check src/prime_rl/dashboard/static/app.js`
- browser-tested Semantic against real RedSearch rollouts on DeepSeek V4 Flash with concurrent subagents and automatic compaction, including long 66- and 80-call traces
- verified against two timeout-interrupted 400+ call RedSearch traces: five stranded final calls rejoined their proven subagent tool flows, leaving exactly one root and three subagents per trace
- verified causal-rank alignment, directional fork/join/compaction edges, fold expansion, call selection, transcript highlighting, context usage, and horizontal scrolling under large fan-out

---

> [!NOTE]
> **Low Risk**
> Changes are read-only dashboard visualization and timeline projection; physical trace data is not mutated, with unit tests for semantic projection edge cases.
> 
> **Overview**
> Adds a **Semantic** tab to the trace viewer (alongside Transcript and Timeline) that visualizes model-call relationships from persisted `MessageNode.semantic_parents`.
> 
> The timeline API now projects **semantic lanes**, **edges**, and **`latest_context_tokens`** in `project_episode_timeline`: continuation components become agent/context lanes; subagent and compaction edges split contexts; missing continuations can be **inferred** from completed tool round-trips or labeled **unlinked** when lineage is ambiguous.
> 
> The frontend renders a causal graph (rank-based layout, bounded agent lanes, foldable ordinary-call runs), a usage inspector, episode prev/next in semantic mode, and **open in transcript** with highlight. Docs and the dashboard skill describe the three views. Unit tests cover fork/join, compaction, inference, and unlinked cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d11cf786b7b91394ba90b1f1ab938af105de6093. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/dashboard/static/app.js [hunk 1]
```diff
@@ -2828,35 +2852,443 @@ function timelineLaneHtml(lane, start, total) {
...
+    const agentEvents = events
+      .filter((event) => event.agent === agent)
+      .sort(
+        (left, right) =>
+          (left.span.started_at ?? Infinity) - (right.span.started_at ?? Infinity) ||
+          left.nodeIndex - right.nodeIndex
+      );
+    for (let index = 1; index < agentEvents.length; index++) {
+      addCausalLink(agentEvents[index - 1].key, agentEvents[index].key);
+    }
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/dashboard/static/app.js:3607]: Fixed in `b2210be11`. Timestamp-ordered same-agent links are now added only when neither causal direction is already reachable in the explicit graph. This keeps the fallback useful for disconnected calls without allowing missing or inverted timestamps to reverse semantic provenance or introduce a cycle.

File: src/prime_rl/dashboard/static/app.js [hunk 2]
```diff
@@ -2828,35 +2855,541 @@ function timelineLaneHtml(lane, start, total) {
...
+
+  const collapseRuns = [];
+  const foldRun = (segment, ordinaryEvents) => {
+    const contextCalls = 2;
+    const minimumOrdinaryCalls = contextCalls * 2 + 3;
+    if (ordinaryEvents.length < minimumOrdinaryCalls) return;
+    const hiddenEvents = ordinaryEvents.slice(contextCalls, -contextCalls);
+    const first = hiddenEvents[0];
+    const last = hiddenEvents[hiddenEvents.length - 1];
+    const key = `${currentLine}:${currentTraceIdx}:${segment.key}:${first.key}:${last.key}`;
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/dashboard/static/app.js]: Fixed in `b2210be11`. Fold keys no longer include `currentTraceIdx`; the episode, semantic segment, and hidden-run endpoints already provide the stable identity. Switching transcript tabs therefore no longer invalidates expanded-run state.

File: src/prime_rl/dashboard/static/app.js [hunk 3]
```diff
@@ -2828,35 +2855,541 @@ function timelineLaneHtml(lane, start, total) {
...
+    .join("");
+
+  target.innerHTML =
+    timelineLegendHtml(edges) +
+    `<div class="sg-shell" style="width:${width}px;height:${height}px">` +
+    laneBands +
+    `<svg class="sg-edge-layer" viewBox="0 0 ${width} ${height}" width="${width}" height="${height}" aria-hidden="true"><defs>${markers}</defs>${lifelines}${paths}</svg>` +
+    agentLabels +
+    turns +
+    runControls +
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/dashboard/static/app.js:3992]: Fixed in `b2210be11`. Same-episode rerenders now capture and restore both `scrollLeft` and `scrollTop`; only a new episode receives initial root centering. The expanded fold control also stays on the lifeline instead of moving laterally. Verified at a constrained 900px viewport: `scrollLeft` remains `128px → 128px → 128px` through expand and re-collapse.

File: src/prime_rl/dashboard/static/app.js [hunk 4]
```diff
@@ -2828,35 +2852,453 @@ function timelineLaneHtml(lane, start, total) {
...
+      const side = Math.sign(segment.agent.offset);
+      const labelDirection = -side;
+      const labelLeft = isRoot
+        ? position.x - labelWidth / 2
+        : position.x + labelDirection * 90 - labelWidth / 2;
+      const labelTop = isRoot ? position.y - 34 : position.y - 30;
+      const sideClass = isRoot ? "root" : labelDirection < 0 ? "side-left" : "side-right";
+      return (
+        `<div class="sg-agent-label ${sideClass} ${segment.incomingType === "compaction" ? "compacted" : ""}" ` +
+        `style="left:${labelLeft}px;top:${labelTop}px;width:${labelWidth}px">` +
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/dashboard/static/app.js:3937]: Addressed by the subsequent geometry refinements, culminating in `b2210be11`. Segment labels now sit above/outside the first call, while call edges terminate at split side ports. I rechecked the dense 80-call trace at 3018×1956: both subagent-call arrowheads and the first-turn labels remain fully visible with no opaque label overlap.

File: src/prime_rl/dashboard/static/app.js [hunk 5]
```diff
@@ -2229,9 +2229,17 @@ let currentBranchIdx = 0;
 let episodeOpenVersion = 0;
 let episodeEnrichmentVersion = 0;
 let currentTimeline = null;
-let traceView = prefs.traceView === "timeline" ? "timeline" : "transcript";
+let traceView = ["timeline", "semantic"].includes(prefs.traceView) ? prefs.traceView : "transcript";
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/dashboard/static/app.js]: Fixed in `df39ed919`. The dashboard now separates the saved preferred view from the effective per-episode view: episodes without semantic data fall back to Transcript while keeping Semantic as the preference, so returning to a compatible episode restores Semantic automatically. Browser-tested across semantic → non-semantic → semantic episodes; the fallback remained readable and the preference was preserved.

File: src/prime_rl/dashboard/static/app.js [hunk 6]
```diff
@@ -3188,35 +3274,686 @@ function timelineLaneHtml(lane, start, total) {
...
+      `<div class="sg-inspector-section">context length</div>` +
+      semanticInspectorRows(semanticContextRows(segment.lane.usage, false)) +
+      `<div class="sg-inspector-section">cumulative usage</div>` +
+      semanticInspectorRows(semanticCumulativeUsageRows(segment.lane.usage, false));
+  }
+  $("#sg-inspector").hidden = false;
 }
 
 async function setTraceView(view) {
+  preferredTraceView = view;
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/dashboard/static/app.js]: Fixed in `59e13ece5`. `setTraceView` now distinguishes persistent tab selection from temporary navigation: **Open in transcript** switches only the effective view, while direct clicks on a different tab still update the saved preference. Clicking the already-active fallback Transcript tab is a no-op. Browser-tested Semantic → Open in transcript → another semantic episode (Semantic restored), plus an explicit Transcript selection across episodes (Transcript persisted); no console errors.

File: src/prime_rl/dashboard/static/app.js [hunk 7]
```diff
@@ -2410,7 +2438,9 @@ async function openEpisode(line, target = {}) {
   currentEpisode = episode;
   currentTraceIdx = target.trace ?? 0;
   currentBranchIdx = target.branch ?? 0;
-  if (traceView === "timeline") await ensureTimeline();
+  traceView = preferredTraceView;
+  if (traceView === "timeline" || traceView === "semantic") await ensureTimeline();
+  if (traceView === "semantic" && !timelineHasSemantic()) traceView = "transcript";
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/dashboard/static/app.js:2451]: Fixed in `23de40d4e`. `openEpisode` now rechecks both line and `episodeOpenVersion` immediately after the awaited timeline fetch, before applying the per-episode Semantic fallback or rendering. A stale open therefore cannot mutate the newer episode view.

File: src/prime_rl/dashboard/static/app.js [hunk 8]
```diff
@@ -2736,6 +2770,84 @@ function toolDefinitionsHtml(trace) {
...
+        `<div class="judge-call-grid"><section class="judge-request"><header>Judge prompt</header>${prompt}</section>` +
+        `<section class="judge-result"><header>Judge response</header><pre>${esc(response.text)}</pre></section>${parsed}</div>` +
+        `</details>`
+      );
+    })
+    .join("");
+  return (
+    `<div class="judging-divider"><span>Judging</span><span>${records.length} call${records.length === 1 ? "" : "s"}</span></div>` +
+    calls
+  );
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/dashboard/static/app.js:2860]: This compatibility path was intentionally removed in the subsequent strict-contract revision. New evidence exists only as required `{name, request, response}` entries under `trace.info["judge_calls"]`; old `trace.info["judge"]` is ignored and therefore never dereferenced or able to break transcript rendering. A malformed `judge_calls` entry is now a producer contract violation, not a supported legacy shape. The PR description and browser QA have been updated accordingly.

File: src/prime_rl/dashboard/static/app.js [hunk 9]
```diff
@@ -3188,35 +3359,686 @@ function timelineLaneHtml(lane, start, total) {
...
+  segments.forEach((segment) => {
+    segment.events.sort((left, right) => left.row - right.row);
+    segment.firstEvent = segment.events[0];
+    segment.lastEvent = segment.events[segment.events.length - 1];
+  });
+  const incomingEdgeByTarget = new Map(
+    edges.map((edge) => [semanticNodeKey(edge.trace_index, edge.target_node), edge])
+  );
+  for (const segment of segments) {
+    segment.incomingType = incomingEdgeByTarget.get(segment.firstEvent?.key)?.type || "";
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/dashboard/static/app.js]: Fixed in `014360b12`. Compaction treatment now comes from a set of all compaction targets, so another return or custom parent cannot overwrite it. Both the compacted label and styling use the resulting `isCompacted` fact; dashboard tests, syntax checks, and pre-commit pass.

File: src/prime_rl/dashboard/static/app.js [hunk 10]
```diff
@@ -3188,35 +3359,686 @@ function timelineLaneHtml(lane, start, total) {
...
+}
+
+function renderSemanticGraph() {
+  const target = $("#tm-timeline");
+  const wasSemantic = target.classList.contains("semantic-canvas");
+  target.classList.add("semantic-canvas");
+  const episodeKey = String(currentLine);
+  const preserveViewport = wasSemantic && target.dataset.semanticEpisode === episodeKey;
+  const previousScrollLeft = target.scrollLeft;
+  const previousScrollTop = target.scrollTop;
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/dashboard/static/app.js:3476]: Fixed in `014360b12`. A fresh `openEpisode` now clears the semantic-canvas class and episode viewport identity before installing the loading state. In-place rerenders still preserve scroll, while closing and reopening—even the same episode—takes the initial centering path and centers the root again.

**Review by samsja — APPROVED** [review]
